### PR TITLE
chore(package.json): Update dependencies and node version

### PR DESCRIPTION
**Pull request by 1p22geo**
**What kind of change does this PR introduce?**

- Use node v20.9.0+ (or anything > 20.9.0) as specified in README instead of stricly requiring node == 20.0.0 (as it was in package.json)
- Update to use latest yarn
- Yarn is a package manager and not a dependency, it is not needed in runtime. Yarn installs the dependencies, so no need to mention it in `dependencies` in package.json

**Issue Number:**
There weren't any, as it is not *technically* a bug. Yet.

**Screenshots/videos:**

Sorry, none here.

**If relevant, did you update the documentation?**

Not relevant.

**Summary**

[why upgrade yarn](https://yarnpkg.com/migration/overview)

I know that `package.json` updates are stricly forbidden for this repository (there is even a CI workflow to fail PRs which changed it) but this is kinda important.

I know it will cause major pain, especially for CI runs, but I can help with migrating workflows to yarn 4.

**Does this PR introduce a breaking change?**

Yes.

### CODE REVIEW

**Comment by 1p22geo** [comment]
Hold on, I am fixing the build errors...

**Comment by 1p22geo** [comment]
Builds and CI succeed now, and the only thing failing is the `unauthorized file edit` workflow. All workflows and builds use latest yarn and node now.

**Comment by benjagm** [comment]
Thanks a lot for this PR! We'll need some extra time to review it but hope to provide feedback tomorrow the latest.

**Comment by 1p22geo** [comment]
Sorry for the tons of unnecessary updates here, I just wanted to fix conflicts with upstream (after merging #729 and #730 which also updated package.json and yarn.lock).

File: .github/workflows/pull-request.yml
Comment (aialok) [inline .github/workflows/pull-request.yml:21]: Is there any advantage of using `yarn workspaces focus -A` over  `yarn install --frozen-lockfile` as we don't have multiple workspaces?

File: .github/workflows/pull-request.yml
Comment (aialok) [inline .github/workflows/pull-request.yml:102]: See comment : https://github.com/json-schema-org/website/pull/727/files#r1630792501

File: package.json
Comment (aialok) [inline package.json:21]: I think it is better to use "node": "^20.0.0" over "node": ">20.9.0" because the latter can install the latest version of Node.js, which might not be stable and could introduce breaking changes. It's better to stick with a stable version. If Node.js warns about compatibility issues or security updates, we can then evaluate and upgrade to a newer stable version as needed.

**Review by aialok — CHANGES_REQUESTED** [review]
See comment. Apart from the comment (LGTM).

File: package.json
Comment (benjagm) [reply package.json:21]: In this case I agree in going with 20.9.0 but fixed, not ">". 20.9.0 is the node version we are running in production in Cloudflare

Screenshot 2024-06-07 at 10 07 01

File: .github/workflows/pull-request.yml
Comment (1p22geo) [reply .github/workflows/pull-request.yml:21]: Quoting the [documentation](https://yarnpkg.com/cli/workspaces/focus),
```
If the -A,--all flag is set, the entire project will be installed. Combine with --production to replicate the old yarn [install](https://yarnpkg.com/cli/install) --production.
```
It's the Yarn 4 equivalent of `yarn install`. But I guess I can restore it.

**Comment by 1p22geo** [comment]
- No more `yarn workspaces focus`, everywhere `yarn install --frozen-lockfile`
- node `^20.9.0` in package.json
- node `20.9.0` in github actions

File: .github/workflows/pull-request.yml
Comment (1p22geo) [reply .github/workflows/pull-request.yml]: Done, `yarn install --frozen-lockfile` everywhere

File: .github/workflows/pull-request.yml
Comment (1p22geo) [reply .github/workflows/pull-request.yml]: alr, `yarn install --frozen-lockfile` back to normal

File: package.json
Comment (1p22geo) [reply package.json]: node `==20.9.0` in package.json and github actions now.

File: package.json
Comment (aialok) [inline package.json:34]: I think we have removed this from our project recently `"next-plausible": "^3.11.1",` .

Related PR : https://github.com/json-schema-org/website/pull/737

Please make sure you are in latest upstream.

File: package.json
Comment (aialok) [inline package.json:52]:
```suggestion
    "@types/node": "^20.9.0",
```

**Comment by aialok** [comment]
@1p22geo  Please make sure you always fetch latest upstream as this is going to be major changes in our codebase.  

Thank you  : )

**Comment by 1p22geo** [comment]
I see the errors, fixing now...

File: package.json
Comment (1p22geo) [reply package.json]: updated to upstream - 6bc564e0d91041a1e3f54babf2a87cbbe100ae61

**Review by aialok — APPROVED** [review]
LGTM :rocket:

**Warning:** This PR might introduce breaking changes in production, as there are complete changes in our `yarn.lock` file and yarn is now in an upgraded version.

**Comment by benjagm** [comment]
Hi @1p22geo do you mind if we change the target branch to try the deployment first in cloudflare? Once that works well merge the temorary branch to main.

**Comment by 1p22geo** [comment]
Updated to upstream and resolved conflicts.

**Comment by 1p22geo** [comment]
BTW may I know how is the whole "temporary deployment" thing going? Are there any errors that need to be fixed, or is it just difficult to set up?

**Review by benjagm — APPROVED** [review]
Thanks for your contribution!

**Comment by benjagm** [comment]
> BTW may I know how is the whole "temporary deployment" thing going? Are there any errors that need to be fixed, or is it just difficult to set up?

We have been busy!! Merged!

**Comment by benjagm** [comment]
I tested everything in cloudflare and I needed to add yarn back as dependency because we are using it to generate the sitemap files at deployment time. You can see the content fo the branch here: https://github.com/json-schema-org/website/tree/web-node-20-9-0

This is the build command we are running in cloudflare :  next build && next-sitemap --config next-sitemap.config.js

Now everything is working in web-node-20-9-0 and we are ready to merge to main.

**Comment by 1p22geo** [comment]
Why is it yarn 1.22 again? I added Yarn 4...

**Comment by 1p22geo** [comment]
packageManager is yarn 4.2 (latest
https://github.com/json-schema-org/website/blob/674ac5d1368e106930803860a478bc56330358fa/package.json#L76
but you added yarn 1.22?
https://github.com/json-schema-org/website/blob/674ac5d1368e106930803860a478bc56330358fa/package.json#L45

**Comment by benjagm** [comment]
Last version of yarn is 1.22

https://github.com/yarnpkg/yarn/releases

**Comment by 1p22geo** [comment]
image
Then am I a time traveller?

**Comment by benjagm** [comment]
:-) I am not an expert. I am just trying to keep the website running while managing the community

**Comment by 1p22geo** [comment]
Also, it's the stale repo.
[https://github.com/yarnpkg/berry](https://github.com/yarnpkg/berry)

**Comment by 1p22geo** [comment]
Ah wait, somehow `yarn@latest` is yarn 1.22, but `yarn@stable` is yarn 4.2
`npm install yarn` or `yarn add yarn` will add yarn 1.22 (default)
`yarn add yarn@stable` or `yarn set version stable` will install ACTUAL latest

**Comment by benjagm** [comment]
It seems than the last yarn version available as dependency is the 1.22.22 version:

https://yarnpkg.com/package?q=yarn&name=yarn

**Comment by 1p22geo** [comment]
Wait, you said you run `next build && next-sitemap --config next-sitemap.config.js` on the Cloudflare worker.

But `next` is not a regular binary file. To run it you need either `yarn` or `npx` or a `package.json` script. Which means that `yarn` is installed already, before running anything. 

`yarn` is the thing that's doing the installing, so it shouldn't really be installing itself.

Also you can run `yarn set version stable` to install yarn v4

**Comment by benjagm** [comment]
Hi again @1p22geo. We are not using cloudflare workers, we are using cloudflare pages. If we remove yarn from the dependencies this is the error we in the cloudflare deployment:

````

13:45:10.141 | Verify run directory
-- | --
13:45:10.142 | Executing user command: next build && next-sitemap --config next-sitemap.config.js
13:45:10.145 | /opt/build/bin/build: line 39: next: command not found
13:45:10.149 | Failed: build command exited with code: 127
13:45:11.023 | Failed: error occurred while running build command

````

I'd like to clarify that we are not expert frontend developers, we are a group o people passionate about JSON Schema, the Specification and its tooling. These type of changes that maybe are trivial for you, may be more complex for us. Our expertise field is the spec, JSON Schemas design and processing and developer tooling .... so please bear with.

**Comment by jdesrosiers** [comment]
I agree that it doesn't make sense for `yarn` to be a dependency. Ideally you would call `yarn run build` instead of `next build && next-sitemap --config next-sitemap.config.js` in Cloudflare Pages. It runs the same commands, but does so in the context of the installed dependencies (like `next` and `next-sitemap`). @benjagm, can we give that a try on Cloudflare Pages without `yarn` in the dependencies and see if that works?

**Comment by jdesrosiers** [comment]
@benjagm I remembered that I have access the Cloudflare Pages, so I tried it myself. I changed the command to `yarn build` and redeployed the branch without the yarn dependency. Everything deployed fine, but it seems Cloudflare Pages decided to use yarn 1.22.4 rather than 4.2.2 as declared in package.json. 

```
17:28:44.100 | Installing yarn at version 1.22.4
```

It seems that it doesn't respect the `packageManager` field. Not sure if there's a way around that problem, but I'd say that's a blocker for upgrading to yarn v4 if that problem can't be solved.

After my testing, I changed all the settings back to how they were before. I'll let you decide whether or not to make the change permanent.

**Comment by jdesrosiers** [comment]
So, I found out that Cloudflare Pages has v1 and v2 build systems. The v2 build system respects `packageManager`, but we were using v1. I did another test deployment, this time using the v2 build system and `yarn bulid` as the build command. Everything seems to have worked and it's using the right version of yarn this time.

```
17:49:54.036 | Preparing yarn@4.2.2 for immediate activation...
```

@benjagm, I reverted all the Cloudflare configuration changes I made so you have a chance to do more thorough testing (I did little more than make sure it built without errors) before making the changes permanent.

**Comment by jdesrosiers** [comment]
@1p22geo I want to thank you for taking the time to do this upgrade. Keeping dependencies up-to-date is a very important security practice.

That said, I want to call your attention to an interaction that occurred in this thread.

> Then am I a time traveller?

> I literally gave you this link in the PR description https://yarnpkg.com/migration/overview

Some of us found these comments to be disrespectful. I know cultural differences as well as language barriers can lead to misunderstandings especially in async written communications like this, but I wanted you to know this was how it was received and I hope you'll adjust as I'm sure you don't mean to offend anyone. @benjagm might not be a web development expert, but this website and many other things in this community wouldn't exist without his efforts. He deserves our respect.

Thanks again for your contributions and I'm glad to have you as part of this community.

**Comment by benjagm** [comment]
Thanks everyone! @jdesrosiers Good finding with the v1 vs v2 build system of cloudflare. I changed it and it starts using the modern Yarn. That works.

I'd like to recall that we are doing all of this use site-map to generate the sitemap xml files and improve our SEO.  These are the scripts we have configured in package.json:

````
  "scripts": {
    "build": "next build",
    "postbuild": "next-sitemap --config next-sitemap.config.js",
````

I have checked the result of using `yarn run build` in cloudflare and unfortunately it doesn't the **postbuild** script necessary to generate the sitemap. After doing some research i found [this question in their forum](https://community.cloudflare.com/t/running-two-build-commands-sequentially-in-build-configurations/416017) and I ended up using `yarn run build && yarn run postbuild` and now everything works as expected:
- Yarn removed from dependencies
- Yarn as package manager with V 4.2.2
- Sitemap generate properly in cloudflare

**Comment by jdesrosiers** [comment]
I'm surprised that `postbuild` doesn't run automatically when you run `build`. I thought that was its purpose. Anyway, that's why I left it for you to verify things were working :smile:.

**Pull request merged**
